include/vara/Feature/Constraint.h

```diff
@@ -37,161 +45,216 @@ class NumericConstraint {
-      : BinaryConstraint(LeftOperand, RightOperand) {}
+  DivisionConstraint(std::unique_ptr<Constraint> LeftOperand,
+                     std::unique_ptr<Constraint> RightOperand)
+      : BinaryConstraint(std::move(LeftOperand), std::move(RightOperand)) {}
 
   [[nodiscard]] std::string toString() const override;
 };
 
 class LessConstraint : public BinaryConstraint, public NumericConstraint {
 public:
-  LessConstraint(Constraint *LeftOperand, Constraint *RightOperand)
-      : BinaryConstraint(LeftOperand, RightOperand) {}
+  LessConstraint(std::unique_ptr<Constraint> LeftOperand,
+                 std::unique_ptr<Constraint> RightOperand)
+      : BinaryConstraint(std::move(LeftOperand), std::move(RightOperand)) {}
 
   [[nodiscard]] std::string toString() const override;
 };
 
 class GreaterConstraint : public BinaryConstraint, public NumericConstraint {
 public:
-  GreaterConstraint(Constraint *LeftOperand, Constraint *RightOperand)
-      : BinaryConstraint(LeftOperand, RightOperand) {}
+  GreaterConstraint(std::unique_ptr<Constraint> LeftOperand,
+                    std::unique_ptr<Constraint> RightOperand)
+      : BinaryConstraint(std::move(LeftOperand), std::move(RightOperand)) {}
 
   [[nodiscard]] std::string toString() const override;
 };
 
 class LessEquelsConstraint : public BinaryConstraint, public NumericConstraint {
 public:
-  LessEquelsConstraint(Constraint *LeftOperand, Constraint *RightOperand)
-      : BinaryConstraint(LeftOperand, RightOperand) {}
+  LessEquelsConstraint(std::unique_ptr<Constraint> LeftOperand,
+                       std::unique_ptr<Constraint> RightOperand)
+      : BinaryConstraint(std::move(LeftOperand), std::move(RightOperand)) {}
 
   [[nodiscard]] std::string toString() const override;
 };
 
-class GreaterEquelsConstraint : public BinaryConstraint, public NumericConstraint {
+class GreaterEquelsConstraint : public BinaryConstraint,
+                                public NumericConstraint {
 public:
-  GreaterEquelsConstraint(Constraint *LeftOperand, Constraint *RightOperand)
-      : BinaryConstraint(LeftOperand, RightOperand) {}
+  GreaterEquelsConstraint(std::unique_ptr<Constraint> LeftOperand,
+                          std::unique_ptr<Constraint> RightOperand)
+      : BinaryConstraint(std::move(LeftOperand), std::move(RightOperand)) {}
 
   [[nodiscard]] std::string toString() const override;
 };
 
+class Feature;
+class FeatureModelBuilder;
+
 class PrimaryConstraint : public Constraint {
 public:
-  PrimaryConstraint(Feature *Feature)
-      : Feature(Feature), Constraint(ConstraintKind::CK_PRIMARY) {}
+  PrimaryConstraint(std::variant<Feature *, std::string> FV)
+      : FV(std::move(FV)), Constraint(ConstraintKind::CK_PRIMARY) {}
+
+  std::variant<Feature *, std::string> getFeature() { return FV; }
+
+  void accept(ConstraintVisitor &V);
 
 private:
-  Feature *Feature;
+  friend FeatureModelBuilder;
+
+  std::variant<Feature *, std::string> FV;
```

Comment:
What is the reasoning behind the string alternative?

Reply:
While parsing  / building the constraint tree, the actual feature might not be present in the FM, so the name is used as placeholder and replaced afterwards. On the other hand we are able to build a constraint tree without requiring any feature parsing.